server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +27,116 @@ import org.http4s.Request
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeTokenAfterRefill = createBucket.flatMap { testee =>
+        for {
+          _ <- testee.takeToken
+          unavailable <- testee.takeToken
+          _ <- IO.sleep(someRefillTime)
+          available <- testee.takeToken
+        } yield unavailable -> available
+      }
 
-    val takeTokenAfterRefill = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(101.milliseconds) *>
-        testee.takeToken
+      TestControl
+        .executeEmbed(takeTokenAfterRefill)
+        .map { case (a, b) => a.isInstanceOf[TokenUnavailable] && b == TokenAvailable }
+        .assert
     }
-
-    takeTokenAfterRefill
-      .map { result =>
-        ctx.advanceAndTick(101.milliseconds)
-        result
-      }
-      .assertEquals(TokenAvailable)
   }
 
   test("LocalTokenBucket should not add another token at specified interval when at capacity") {
-    val ctx = TestContext()
-    val capacity = 5
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
-
-    val takeExtraToken = createBucket.flatMap { testee =>
-      val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
-        testee.takeToken
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeExtraToken = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
+          testee.takeToken
+        }
+        takeFiveTokens >> IO.sleep(someRefillTime - 1.milliseconds) >> testee.takeToken
       }
-      IO.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
-    }
 
-    takeExtraToken
-      .map { result =>
-        ctx.advanceAndTick(300.milliseconds)
-        result
-      }
-      .map(_.isInstanceOf[TokenUnavailable])
-      .assert
+      TestControl
+        .executeEmbed(takeExtraToken)
+        .map {
+          case TokenUnavailable(Some(_)) => true
+          case _ => false
+        }
+        .assert
+    }
   }
 
   test(
     "LocalTokenBucket should only return a single token when only one token available and there are multiple concurrent requests"
   ) {
     val capacity = 1
     val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+      TokenBucket.local[IO](capacity, 1.milliseconds)
 
     val takeTokensSimultaneously = createBucket.flatMap { testee =>
       (1 to 5).toList.parTraverse(_ => testee.takeToken)
     }
 
-    takeTokensSimultaneously
+    val prog = takeTokensSimultaneously
       .map { result =>
         result.count(_ == TokenAvailable)
       }
-      .assertEquals(1)
+
+    TestControl.executeEmbed(prog).assertEquals(1)
   }
 
+  val localGen = for {
+    fd1 <- genFiniteDuration
+    fd2 <- genFiniteDuration.suchThat(_ < fd1)
+  } yield fd1 -> fd2
```

Comment:
Would it be more efficient to generate two finite durations and sum them to get a monotonically increasing sequence?

Reply:
I'm not sure I'm seeing the improved efficiency - may be the Saturday morning brain slowness :grinning: care explaining how you'd do it and why it'd be more efficient with an increasing sequence rather than working out the difference between the 2 FDs?

Reply:
IIUC the problem is that `genFiniteDuration.suchThat(_ < fd1)` is a rejection sampler: it will keep generating new samples until it satisfies the criterion that `fd2 < fd1`.

Instead, what if we just do:
```scala
val localGen = for {
  fd2 <- genFiniteDuration
  delta <- genFiniteDuration
  fd1 = fd2 + delta
} yield fd1 -> fd2
```

Reply:
Another alternative could be:
```scala
val localGen = for {
  fd1 <- genFiniteDuration
  fd2 <- genFiniteDuration
} yield fd1.max(fd2) -> fd1.min(fd2)
```

Reply:
Ohh, I see what you mean by "adding" the FDs! 
Yes `suchThat` is indeed a rejection sampler, and in some scenarios it could lead to the exhaustion of the max allowed attempts and cause a spurious test failure. The problem that I see with both your suggestions is that the sampler could pick zero as value for both the samples, resulting in `0 -> 0` as result, which would invalidate the semantics of the test, I believe: in fact, if we allow a wait time of 0 how do we quantify that the fiber _has to wait_ in order to pick up the next available token? Now that I'm thinking of it, perhaps we should also enforce that the refill time (a.k.a: `fd1`) should be != 0?

Reply:
Good point about `0`. One lazy solution is to just `.map(_ + 1.millis))` as needed 😉 (`TestControl` has millisecond-resolution).

> Now that I'm thinking of it, perhaps we should also enforce that the refill time (a.k.a: `fd1`) should be != 0?

Yep, might be good to add a `require` in `TokenBucket.local`. Currently seems like it will lead to an `NaN` via divide-by-zero.
https://github.com/http4s/http4s/blob/32d0c4644144234ccc307b5902e91d5015e22284/server/shared/src/main/scala/org/http4s/server/middleware/Throttle.scala#L71

Reply:
yeah, I'll go for the lazy solution on the generator :-)

Reply:
I also took the chance to revisit a bit the code for `TokenBucket.local`